File: go/parsers/unsubscribed.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Unsubscribed struct {
	gen *gen.Unsubscribed
}

func NewUnsubscribedFields(g *gen.Unsubscribed) messages.UnsubscribedFields {
	return &Unsubscribed{gen: g}
}

func (u *Unsubscribed) RequestID() uint64 {
	return u.gen.RequestID()
}

func UnsubscribedToCapnproto(m *messages.Unsubscribed) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	unsubscribed, err := gen.NewRootUnsubscribed(seg)
	if err != nil {
		return nil, err
	}

	unsubscribed.SetRequestID(m.RequestID())

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeUnsubscribed, data, nil), nil
}

func CapnprotoToUnsubscribed(data []byte) (*messages.Unsubscribed, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	unsubscribed, err := gen.ReadRootUnsubscribed(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewUnsubscribedWithFields(NewUnsubscribedFields(&unsubscribed)), nil
}

File: python/wamp_msgs_capnp/parsers/unsubscribed.py
import os
from pathlib import Path

import capnp
from wampproto.messages.unsubscribed import Unsubscribed, IUnsubscribedFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "unsubscribed.capnp")
unsubscribed_capnp = capnp.load(str(schema_file))


class UnsubscribedFields(IUnsubscribedFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID


def unsubscribed_to_capnproto(u: Unsubscribed) -> bytes:
    unsubscribed = unsubscribed_capnp.Unsubscribed.new_message()
    unsubscribed.requestID = u.request_id

    packed_data = unsubscribed.to_bytes_packed()
    return helpers.prepend_header(Unsubscribed.TYPE, packed_data, b"")


def capnproto_to_unsubscribed(data: bytes) -> Unsubscribed:
    message_data, _ = helpers.extract_message(data)
    unsubscribed_obj = unsubscribed_capnp.Unsubscribed.from_bytes_packed(message_data)

    return Unsubscribed(UnsubscribedFields(unsubscribed_obj))
